### PR TITLE
1156 use more specific class names when adding to host element

### PR DESCRIPTION
**Pull request by GCHQ-Developer-530**
## Summary of the changes
Update components with class names on host to be more specific by including the component name in
the class name

BREAKING CHANGE: Class names that are on the host for most component have been updated to include
the component name in the class name e.g. disabled updated to ic-select-disabled. Components updated
are: accordion, accordion group, badge, breadcrumb, breadcrumb group, button, checkbox, checkbox
group, classification banner, data entity, data row, dialog, empty state, footer, hero, horizontal
scroll, input component container, input label, input validation, link, loading indicator, menu,
navigation group, page header, popover menu, radio group, radio option, search bar, select, side
nav, skeleton, step, stepper, tab, tab group, text field, toast, toggle button, toggle button group,
and typography

## Related issue
#1156 

### General 

- [x] Changes to docs package checked and committed.
- [x] All acceptance criteria reviewed and met. 

### Testing

- [x] Relevant unit tests and visual regression tests added. 
- [x] Visual testing against Figma component specification completed. 
- [x] Playground stories in React Storybook up to date, with any prop changes and additions addressed. 

### Accessibility 

- [x] Accessibility Insights FastPass performed.
- [x] A11y unit test added and yields no issues.
- [x] A11y plug-in on Storybook yields no issues. 
- [x] Manual screen reader testing performed using NVDA and VoiceOver. 
- [x] Manual keyboard testing for keyboard controls and logical focus order. 
- [x] Correct roles used and ARIA attributes used correctly where required. 
- [x] Logical heading structure is maintained, and the HTML elements used for headings can be changed to fit within the wider page structure. 

### Resize/zoom behaviour 

- [x] Page can be zoomed to 400% with no loss of content. 
- [x] Screen magnifier used with no issues. 
- [x] Text resized to 200% with no loss of content.
- [x] Text spacing increased as per the [WCAG 1.4.12 success criterion](https://www.w3.org/TR/WCAG21/#text-spacing) with no loss of content.

### System modes

- [x] Browser setting 'prefers reduced motion' tested. No animations or motion visible whilst this setting is on. 
- [x] Windows High Contrast mode tested with no loss of content.
- [x] System light and dark mode tested with no loss of content.
- [x] Browser support tested (Chrome, Safari, Firefox and Edge). 

### Testing content extremes

- [x] Min/max content examples tested with no loss of content or overflow. 
- [x] All prop combinations work without issue. 
- [x] Tested for FOUC (Flash of Unstyled Content) in both SSR (Server-Side Rendering) and SSG (Static Site Generation) settings.
- [x] Controlled and uncontrolled input components tested.

### CODE REVIEW

File: packages/canary-web-components/src/components/ic-date-picker/ic-date-picker.tsx [hunk 1]
```diff
@@ -1313,7 +1313,10 @@ export class DatePicker {
     }
 
     return (
-      <Host onKeyDown={this.keyDownHandler} class={size}>
+      <Host
+        onKeyDown={this.keyDownHandler}
+        class={{ [`ic-date-picker-${size}`]: true }}
```
Comment (ad9242) [inline packages/canary-web-components/src/components/ic-date-picker/ic-date-picker.tsx:1318]:
```suggestion
        class={`ic-date-picker-${size}`}
```

slightly more succinct for just a single class

File: packages/canary-web-components/src/components/ic-menu-with-multi/ic-menu.tsx [hunk 2]
```diff
@@ -1394,16 +1394,16 @@ export class Menu {
     return (
       <Host
         class={{
-          "full-width": fullWidth,
+          "ic-menu-full-width": fullWidth,
           "no-focus":
```
Comment (ad9242) [inline packages/canary-web-components/src/components/ic-menu-with-multi/ic-menu.tsx:1398]: should maybe change "no-focus" too, as still quite a generic name that might clash?

File: packages/canary-web-components/src/components/ic-select-with-multi/ic-select.tsx [hunk 3]
```diff
@@ -1140,11 +1140,11 @@ export class Select {
     return (
       <Host
         class={{
-          disabled: disabled,
+          ["ic-select-disabled"]: disabled,
           searchable: searchable,
```
Comment (ad9242) [inline packages/canary-web-components/src/components/ic-select-with-multi/ic-select.tsx:1144]: searchable too?

File: packages/web-components/src/components/ic-accordion-group/ic-accordion-group.tsx [hunk 4]
```diff
@@ -159,8 +159,8 @@ export class AccordionGroup {
       <Host
         context-id={this.accordionGroupId}
         class={{
-          [`${appearance}`]: true,
-          [`${size}`]: true,
+          [`ic-accordion-group-${appearance}`]: true,
+          [`ic-accordion-group-${size}`]: true,
           ["accordion-group"]: true,
```
Comment (ad9242) [inline packages/web-components/src/components/ic-accordion-group/ic-accordion-group.tsx:164]:
```suggestion
          ["ic-accordion-group"]: true,
```
?

File: packages/web-components/src/components/ic-badge/ic-badge.tsx [hunk 5]
```diff
@@ -235,11 +235,12 @@ export class Badge {
...
-          [`${variant}`]: true,
-          [`${type}`]: true,
-          [`foreground-${foregroundColour}`]: foregroundColour !== null,
+          [`ic-badge-${position}`]: true,
+          [`ic-badge-${size}`]: true,
+          [`ic-badge-${variant}`]: true,
+          [`ic-badge-${type}`]: true,
+          [`ic-badge-foreground-${foregroundColour}`]:
+            foregroundColour !== null,
           [`${visible ? "show" : "hide"}`]: true,
```
Comment (ad9242) [inline packages/web-components/src/components/ic-badge/ic-badge.tsx:244]: should probably do "show" and "hide" to as they are quite generic

File: packages/web-components/src/components/ic-footer/ic-footer.tsx [hunk 6]
```diff
@@ -152,9 +152,9 @@ export class Footer {
           [`footer-${groupLinks ? "grouped" : "ungrouped"}`]: true,
           [`footer-${foregroundColor}`]: true,
           // Slots will be able to infer their own color
-          [IcThemeForegroundEnum.Dark]:
+          [`ic-footer-${IcThemeForegroundEnum.Dark}`]:
```
Comment (ad9242) [inline packages/web-components/src/components/ic-footer/ic-footer.tsx:155]: change "footer" on line 150 too?

File: packages/web-components/src/components/ic-link/ic-link.tsx [hunk 7]
```diff
@@ -149,7 +149,7 @@ export class Link {
     } = this;
 
     return (
-      <Host class={{ ["link"]: true, [`${appearance}`]: true }}>
+      <Host class={{ ["link"]: true, [`ic-link-${appearance}`]: true }}>
```
Comment (ad9242) [inline packages/web-components/src/components/ic-link/ic-link.tsx:152]:
```suggestion
      <Host class={{ ["ic-link"]: true, [`ic-link-${appearance}`]: true }}>
```

File: packages/web-components/src/components/ic-menu/ic-menu.tsx [hunk 8]
```diff
@@ -995,11 +995,11 @@ export class Menu {
     return (
       <Host
         class={{
-          "full-width": fullWidth,
+          "ic-menu-full-width": fullWidth,
           "no-focus": inputEl?.tagName === "INPUT" || hasTimedOut || isLoading,
```
Comment (ad9242) [inline packages/web-components/src/components/ic-menu/ic-menu.tsx:999]: no-focus too?

File: packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx [hunk 9]
```diff
@@ -846,9 +846,9 @@ export class SearchBar {
       <Host
         class={{
           ["search"]: true,
```
Comment (ad9242) [inline packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx:848]: search too?

File: packages/web-components/src/components/ic-select/ic-select.tsx [hunk 10]
```diff
@@ -1050,11 +1050,11 @@ export class Select {
     return (
       <Host
         class={{
-          disabled: disabled,
+          ["ic-select-disabled"]: disabled,
           searchable: searchable,
```
Comment (ad9242) [inline packages/web-components/src/components/ic-select/ic-select.tsx:1054]:
```suggestion
          ic-searchable: searchable,
```

File: packages/web-components/src/components/ic-stepper/ic-stepper.tsx [hunk 11]
```diff
@@ -295,7 +295,7 @@ export class Stepper {
     return (
       <Host
         class={{
-          [`${this.variant}`]: true,
+          [`ic-stepper-${this.variant}`]: true,
           ["aligned-left"]:
```
Comment (ad9242) [inline packages/web-components/src/components/ic-stepper/ic-stepper.tsx:299]:
```suggestion
          ["ic-aligned-left"]:
```

**Review by ad9242 — CHANGES_REQUESTED** [review]
only other comment would be whether to consider just having "ic-disabled", "ic-open", "ic-full-width" rather than "<ic-component-name-disabled" etc. just slightly more succinct

**Comment by GCHQ-Developer-530** [comment]
> only other comment would be whether to consider just having "ic-disabled", "ic-open", "ic-full-width" rather than "<ic-component-name-disabled" etc. just slightly more succinct

Will make the updates for your requested changes. I went with "ic-component-name-" preceding the class name because I noticed a few components were relying on styles from different components due to the generic names, so I figured this way it would stop that and just using "ic-" preceding it wouldn't.

File: packages/canary-web-components/src/components/ic-date-picker/ic-date-picker.tsx (hunk 1 above)
Comment (GCHQ-Developer-530) [reply packages/canary-web-components/src/components/ic-date-picker/ic-date-picker.tsx:1318]: It comes up with "property assignment expected" for this line if I remove the square brackets. Might be something to do with the back ticks

File: packages/web-components/src/components/ic-back-to-top/test/basic/__snapshots__/ic-back-to-top.spec.tsx.snap [hunk 12]
```diff
@@ -32,7 +32,7 @@ exports[`ic-back-to-top should render with an icon only: should render with an i
 `;
 
 exports[`ic-back-to-top should render with footer: should render with footer 1`] = `
-<ic-footer class="footer footer-light footer-sparse footer-ungrouped light">
+<ic-footer class="footer-light footer-sparse footer-ungrouped ic-footer ic-footer-light">
```
Comment (GCHQ-Developer-299) [inline packages/web-components/src/components/ic-back-to-top/test/basic/__snapshots__/ic-back-to-top.spec.tsx.snap:35]: should the 'footer-light' class still exist when 'ic-footer-light' has been created?

Comment (GCHQ-Developer-530) [reply packages/web-components/src/components/ic-back-to-top/test/basic/__snapshots__/ic-back-to-top.spec.tsx.snap]: footer-light was to do with the foreground colour, so I've updated it to "ic-footer-foreground-light" now

File: packages/canary-web-components/src/components/ic-select-with-multi/ic-select.tsx [hunk 13]
```diff
@@ -1140,11 +1140,11 @@ export class Select {
...
-          disabled: disabled,
-          searchable: searchable,
-          small: small,
-          [size]: size !== "default",
-          "full-width": fullWidth,
+          ["ic-select-disabled"]: disabled,
+          ["ic-select-searchable"]: searchable,
+          ["ic-select-small"]: small,
+          [`ic-select-${size}`]: size !== "default",
+          "ic-select-full-width": fullWidth,
```
Comment (ad9242) [inline packages/canary-web-components/src/components/ic-select-with-multi/ic-select.tsx:1147]:
```suggestion
          ["ic-select-full-width"]: fullWidth,
```

just be good to standardize on use of square brackets (or not) for string values - just to match others on lines above

**Review by ad9242 — CHANGES_REQUESTED** [review]

File: packages/web-components/src/components/ic-button/ic-button.tsx [hunk 14]
```diff
@@ -553,13 +553,13 @@ export class Button {
     return (
       <Host
         class={{
-          ["disabled"]: this.disabled && !this.loading,
+          ["ic-button-disabled"]: this.disabled && !this.loading,
           [`button-variant-${this.variant}`]: true,
           [`button-size-${this.size}`]: true,
```
Comment (GCHQ-Developer-299) [inline packages/web-components/src/components/ic-button/ic-button.tsx:558]: button variant and size haven't been prefixed, what's the thinking there?

Comment (GCHQ-Developer-530) [reply packages/web-components/src/components/ic-button/ic-button.tsx:558]: I was just updating ones that were generic, but I do agree these should be prefixed with ic so it matches the others. Will update

**Review by GCHQ-Developer-299 — APPROVED** [review]

**Review by ad9242 — APPROVED** [review]

**Pull request merged**